### PR TITLE
[3.x] Add full support for Android scoped storage.

### LINKED ISSUE #38913: Add support for Android scoped storage requirements
**Godot version:**
3.2.2 Beta 2
**OS/device including version:**
Android
**Issue description:**
Try to save image, get error "No permission grants found for com.google.android.apps.photos"

Read/Write external storage permssions are enabled in Android Export options.  The same code worked in 3.2.1 stable.
**Steps to reproduce:**

**Minimal reproduction project:**
Logcat:
05-20 21:35:58.903 29344 29554 I godot   : Loaded Strokes Drawn
05-20 21:35:57.777  1423  4285 W UriGrantsManagerService: No permission grants found for com.google.android.apps.photos
05-20 21:35:59.052  1423  4285 D ConnectivityService: filterNetworkStateForUid() uid: 10209, pid: 18529
05-20 21:35:59.052  1423  4285 D ConnectivityService: filterNetworkStateForUid() uid: 10209, pid: 18529, networkInfo: [type: WIFI[], state: DISCONNECTED/BLOCKED, reason: (unspecified), extra: (none), failover: false, available: true, roaming: false]
05-20 21:35:59.052  1423  4285 D ConnectivityService: filterNetworkStateForUid() uid: 10209, pid: 18529
05-20 21:35:59.053  1423  4285 D ConnectivityService: filterNetworkStateForUid() uid: 10209, pid: 18529, networkInfo: [type: WIFI[], state: DISCONNECTED/BLOCKED, reason: (unspecified), extra: (none), failover: false, available: true, roaming: false]
05-20 21:35:59.137  1423  4285 W UriGrantsManagerService: No permission grants found for com.google.android.apps.photos
05-20 21:35:59.137  1423  4937 W UriGrantsManagerService: No permission grants found for com.google.android.apps.photos
05-20 21:35:59.195 29344 29554 E godot   : **ERROR**: Can't save PNG at path: '/storage/emulated/0/Pictures/HEAVYPAINT_EXPORT/asd.png'.
05-20 21:35:59.195 29344 29554 E godot   :    At: drivers/png/resource_saver_png.cpp:58:save_image() - Condition "err" is true. Returned: err

### LINKED ISSUE #45467: Image.save_png("user://image_name.png") does not save image in reachable location in Android export
**Godot version:**

3.2.3

**OS/device including version:**

Pixel 2, Android 9

**Issue description:**

Tried saving an image in the user directory with Image.save_png(). It didn't work. I enabled read/write to external storage perms and asked the user to grant them (the popup appeared). I printed the error code and it was 0. I went into my files but I don't see anything that looks like the image I saved.

Edit: Found the file with adb, but I'm not sure how to get to it by using my phone. I want the images to be saved in a place that the user can easily access them.

**Steps to reproduce:**
Get the data of the texture of the default icon and save it with Image.save_png().

**Minimal reproduction project:**

[Test.zip](https://github.com/godotengine/godot/files/5870977/Test.zip)

### LINKED ISSUE #48636: Android: Changes needed to support API 30 (Android 11) requirements from August 2021
**Godot version:**
All Godot versions, but realistically only `master`, `3.x` and `3.3`.
Older versions don't support AABs which will become a requirement. (This might finally force EOL for `2.1`, unless someone is really motivated to backport all this work to the `2.1` branch.)

**Issue description:**

Like clockwork, Google Play has updated requirements for new apps and updates that will be enforced from August 2021 (new apps) and November 2021 (updates). These requirements are described in https://developer.android.com/distribute/best-practices/develop/target-sdk ([archive.org link](https://web.archive.org/web/20210501061433/https://developer.android.com/distribute/best-practices/develop/target-sdk)).

From the above documentation, we need to support:

- [x] Publish with the Android App Bundle format. **Supported in 3.3 and later.**
  - [x] AAB should likely be made the default option, and APK support deprecated? I'd keep it in `3.3` but we can possibly remove it in `3.4` (or deprecate and remove in `3.5` once Google Play no longer accepts APKs).
- [x] Use [Play Asset Delivery](https://developer.android.com/guide/playcore/asset-delivery) or [Play Feature Delivery](https://developer.android.com/guide/playcore/feature-delivery) to deliver assets or features that exceed download size of 150MB. Expansion files (OBBs) will no longer be supported for new apps.
  - [x] Where do we stand here for asset delivery features?
- [x] Target API level 30 (Android 11) or above and adjust for [behavioral changes](https://developer.android.com/about/versions/11/behavior-changes-11). Behavioral changes are:
  - [x] Scoped storage enforcement: Access into external storage directories is limited to an app-specific directory and specific types of media that the app has created. #38913
  - [x] Permissions auto-reset: If users haven't interacted with an app for a few months, the system auto-resets the app's sensitive permissions.
  - [x] Background location access: Users must be directed to system settings in order to grant the background location permission to apps.
  - [x] Package visibility: When an app queries for the list of installed apps on the device, the returned list is filtered.

We need to assess all of the above, implement it ASAP for the `3.x` and `master` branches, and cherry-pick to `3.3`. It would be great if we could have this in a `3.3` release at least by late June/July.

@godotengine/android

### PR DESCRIPTION
**Pull request by m4gr3d**
This was done by refactoring directory and file access handling for the Android platform so that any general filesystem access type go through the Android layer.

This allows us to validate whether the access is unrestricted, or whether it falls under scoped storage and thus act appropriately.

Fix #38913
Fix #45467
Fix #48636

### CODE REVIEW

**Comment by m4gr3d** [comment]
@HEAVYPOLY I fixed the issues causing the crash and rebased the PR.

I did a quick test, and starting and loading a Godot app seems to work as expected. I still need to fully validate scoped storage, and I'm working to complete that later this week.

**Comment by HEAVYPOLY** [comment]
@m4gr3d thanks!  Look forward to testing, this is huge!

**Comment by ondesic** [comment]
Please let me know when this can be tested.  I am excited to try this as well.

**Comment by HEAVYPOLY** [comment]
@m4gr3d I've cherry-picked this commit into my branch of godot and trying to test, how should I set up my project to use these new scoped storage features?

So far I get no crash, saves are going to data/data/com.heaypoly.heavypaint/files/heavypaint_exports/filename.png but I'm still not able to see it in my phones files directory.

**Comment by HEAVYPOLY** [comment]
I've tried using OS.get_system_dir(OS.SYSTEM_DIR_DOCUMENTS) and OS.SYSTEM_DIR_PICTURES, but still not able to save files in android 11.  Is there some other directory I can save to?

**Comment by ondesic** [comment]
@HEAVYPOLY,
Which version are you using?

**Comment by HEAVYPOLY** [comment]
Android 11 , Godot 3.x with this cherry picked

FYI, path.get_file() seems to cause crash on android 11

**Comment by ondesic** [comment]
There are two problems I have ran into.  One is when I go on my phone into permissions to allow storage access, it only has "Allow access to media only" as the only option.  I notice that in the permissions for, say Acrobat reader, it has 2 options. "Allow management of all files" and "Allow access to media only".  How do we get Godot to "Allow management of all files"?

Also, even though it is allowing me to save a file to the documents folder, if I REINSTALL my app, it will not let me access those saved files anymore.  The files are still there, I just can't find them in the FileDialog.

**Comment by m4gr3d** [comment]
> Just want to confirm OS.get_system_dir(OS.SYSTEM_DIR_DOCUMENTS) works in Android 11! SYSTEM_DIR_PICTURES and SYSTEM_DIR_DCIM does not, but I'm so happy there is a way to save files again!

With scoped storage in effects, `Download` and `Documents` are the two most accessible directories that are not app specific, so I'd recommend using the `Documents` directory for data you want to make easily accessible to the user.
 
> It seems make_dir doesn't work inside documents folder, is there a workaround or not implemented yet?

@HEAVYPOLY @ondesic There were a few bugs in the PR with directory access and creation, they should be all fixed now!


> There are two problems I have ran into. One is when I go on my phone into permissions to allow storage access, it only has "Allow access to media only" as the only option. I notice that in the permissions for, say Acrobat reader, it has 2 options. "Allow management of all files" and "Allow access to media only". How do we get Godot to "Allow management of all files"?
> 
> Also, even though it is allowing me to save a file to the documents folder, if I REINSTALL my app, it will not let me access those saved files anymore. The files are still there, I just can't find them in the FileDialog.

The latter is a restriction with scoped storage. If the user upgrades your app, then the app should still have access to the saved files. If the app is uninstalled and reinstalled on the other hand, access to the saved files in the shared storage is lost.

There are a couple of options to address that:
- You can save files in the app specific directory. On uninstall, you can specify that these files remain so the app can access them when it's reinstalled next. The issue with this approach is that the app specific directories are not user-friendly (e.g: `/sdcard/Android/data/<app_package_name>/...`) so if frequent user access is needed / expected, that may be an issue.
- You can request the [`MANAGE_EXTERNAL_STORAGE`](https://developer.android.com/training/data-storage/manage-all-files) permission (added in this PR). This permission, when granted by the user, gives access to all files. The caveat is that capability is considered `dangerous` and the [play store heavily restricts access to that permission to specific use-cases](https://developer.android.com/training/data-storage/manage-all-files#all-files-access-google-play).

**Comment by m4gr3d** [comment]
@akien-mga I fixed the remaining issues in this PR, and I think it should be ready for review.

I've tested the implementation on my side, but I'd appreciate further testing to validate the implementation.

**Comment by m4gr3d** [comment]
> @m4gr3d do I need to compile the version myself, or is there one I can get already compiled?

@ondesic Do you need a build of the Godot Editor, or just of the Godot Android libraries?

**Comment by ondesic** [comment]
@m4gr3d The project I am using for a test is a Mono project.  Would it be possible to get a mono editor build with the export templates to test this?

File: platform/android/java/lib/src/org/godotengine/godot/io/directory/AssetsDirectoryAccess.kt [hunk 1]
```diff
@@ -0,0 +1,178 @@
+/*************************************************************************/
+/*  AssetsDirectoryAccess.kt                                             */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2021 Juan Linietsky, Ariel Manzur.                 */
```
Comment (akien-mga) [inline platform/android/java/lib/src/org/godotengine/godot/io/directory/AssetsDirectoryAccess.kt:8]: Nitpick but there are still 2021 copyright headers.

Comment (m4gr3d) [reply platform/android/java/lib/src/org/godotengine/godot/io/directory/AssetsDirectoryAccess.kt]: Fixed!

**Comment by akien-mga** [comment]
Needs a rebase after cherry-pick of #62297 (which in retrospective could also have been merged only for `3.4`, but well just in case we don't get to merge this in time for 3.5 :)).

**Comment by akien-mga** [comment]
I made a test build with the official buildsystem upgraded to use NDK r23 (#61691) and including this PR.

Could you test @HEAVYPOLY @ondesic @m4gr3d? There's both Android templates and the desktop editor builds as you need the same version to export.
https://downloads.tuxfamily.org/godotengine/testing/3.5-rc-android-scoped%2bndk23/

If I get some confirmations that this work as intended quickly, then we should be able to include it in 3.5-stable too (first in 3.5 RC 5).

**Comment by akien-mga** [comment]
@m4gr3d I've got a report that audio doesn't seem to work in my test build, while it works in 3.5 RC 4: https://twitter.com/MagellanicGames/status/1540251677772730368

**Comment by MagellanicGames** [comment]
Hi, I was asked to provide a repro for the audio bug.  I've tested this in the build @akien-mga provided and it will play the sound effect upon pressing the button on desktop but not on the Android device.

The android device is a Samsung Galaxy s6, running Android v7.0.

I found the bug in my game as the code is used for triggering effects for all manner of entites, whether they be UI, enemies etc.

Apologies that it's just smashed together, it is ripped from my production code and didn't have long to get it done.  
This is code (for better or worse) that I have used since Godot v3.0.

Hope it is of use.

[audio_bug.zip](https://github.com/godotengine/godot/files/8978686/audio_bug.zip)

**Comment by akien-mga** [comment]
I can confirm the regression with @MagellanicGames' reproduction project.

The `Audio.gd` script has some custom logic to parse the files in `res://Audio/Sfx/` (which when exported are only `.import` files) to infer the names of the original WAV files which can then be `load()`ed. That doesn't seem to work with this PR.
Loading with `load("res://Audio/Sfx/item_pickup.wav")` works fine however.

It seems like it boils down to `Directory.open()` not working for `res://` paths on Android.
This minimal project fails too (from the `Directory` docs):
```gdscript
func dir_contents(path):
	var contents = []
	var dir = Directory.new()
	if dir.open(path) == OK:  # Fails here on Android with this PR.
		dir.list_dir_begin()
		var file_name = dir.get_next()
		while file_name != "":
			if dir.current_is_dir():
				print("Found directory: " + file_name)
			else:
				print("Found file: " + file_name)
			contents.append(file_name)
			file_name = dir.get_next()
	else:
		print("An error occurred when trying to access the path.")
	return contents


func _ready():
	# Should find ".." and "." directories, "hello" and "world" files.
	# On Android, fails opening the folder so doesn't find anything.
	$TextEdit.text = str(dir_contents("res://somedir"))
```

MRP: [bug_scoped_storage_open.zip](https://github.com/godotengine/godot/files/8981769/bug_scoped_storage_open.zip)

**Comment by m4gr3d** [comment]
@akien-mga @MagellanicGames Thanks for the repro instructions! 

I found and fixed the issue, so this should be good to go now!

**Review by akien-mga — APPROVED** [review]
Looks great to me!

Tested on Poco F1 (MIUI 12 based on Android 10), I could read from `res://`, write to and read from `user://`, and write to and read from Documents (using `Manage Documents`). Didn't test external storage.

**Comment by akien-mga** [comment]
Thanks!

**Comment by ondesic** [comment]
Just tested my mono project and it is crashing.  Says "Permission to access file: /storage/emulated/0/Documents/Data/Custom.txt is denied"

Also, I changed my Target SDK to 32 to allow RC_5 to run.

**Comment by ondesic** [comment]
Here are my findings with the included test mono project.
This app is coded to take a file stored in "res://Test/Test.txt" and create a folder named "Test" in the Document folder, then copy the Test.txt file into that folder.  When you press the big button, it will read the Test.txt file and assign the string in the file to the button's text.
This runs as expected in Windows.  However, on the android device:
1) if I select "Manage Documents" only, the app will create a "Test" folder but will NOT copy the Test.txt file.  I end up with an empty folder.
2) Next, I uninstall the app and delete the Test folder.  I now add "Manage External Storage".  The same thing happens, only the Test folder is created with no file inside.
3) Next I uninstalled and unchecked all permissions.  I then only checked "Read External Storage" and "Write External Storage".  I got the same results...Test folder created but empty and unable to copy Test.txt to the folder.
4) Finally, No combination worked except checking all of the following "Manage External Storage", "Read External Storage" and "Write External Storage".  Then I have to allow the permissions for "All Files" and then it will finally allow me to copy the Test.txt file..........except...... it makes Test.txt a 0 byte file with nothing inside.
Here is the project.
[Test.zip](https://github.com/godotengine/godot/files/9005370/Test.zip)
.

**Pull request merged**